lib/libzfs/libzfs_pool.c

```diff
@@ -3854,6 +3854,10 @@ zpool_history_unpack(char *buf, uint64_t bytes_read, uint64_t *leftover,
 		if (ISP2(*numrecords + 1)) {
 			*records = realloc(*records,
 			    *numrecords * 2 * sizeof (nvlist_t *));
+			if (*records == NULL) {
```

Comment:
@gmelikov The error code is fine. I was thinking something like the following:
```
		/* Increase the size of our array if necessary */
		if (ISP2(*numrecords + 2)) {
			/* realloc can fail and original memory won't be freed */
			tmp = realloc(*records,
			    (*numrecords + 1) * 2 * sizeof (nvlist_t *));
			if (tmp == NULL) {
				nvlist_free(nv);
				return (ENOMEM);
			}

			*records = tmp;
		}

 		/* add record to nvlist array */
		(*records)[*numrecords] = nv;
		(*numrecords)++;
```

You may want to consider writing comments at the top 'zpool_history_unpack' that describe the contract between this function and the caller. For example, in the snippet above, I am guaranteeing that 'zpool_history_unpack' will not change any of the original values in the scenario of a failure.

Thoughts?